tests/unit/set.cpp

```diff
@@ -51,107 +57,155 @@ static grb::RC dense_tests(
 
 	std::cerr << "\b 2:";
 	ret = grb::set< descriptors::dense >( dst, 1.0 );
+	ret = ret ? ret : grb::wait( dst );
 	if( ret != ILLEGAL ) {
 		std::cerr << " expected ILLEGAL, got " << toString( ret ) << "\n";
 		return FAILED;
 	}
-	if( nnz( dst ) != 0 ) {
-		std::cerr << " expected 0, got " << nnz( dst ) << "\n";
-		return FAILED;
+	if( nonblocking_execution ) {
+		ret = ret ? ret : grb::clear( dst );
+	} else {
+		if( nnz( dst ) != 0 ) {
+			std::cerr << " expected 0, got " << nnz( dst ) << "\n";
+			return FAILED;
+		}
 	}
 
 	std::cerr << "\b 3:";
 	ret = grb::set< descriptors::dense >( dst, one_mask, 1.0 );
+	ret = ret ? ret : grb::wait( dst );
 	if( ret != ILLEGAL ) {
 		std::cerr << " expected ILLEGAL, got " << toString( ret ) << "\n";
 		return FAILED;
 	}
-	if( nnz( dst ) != 0 ) {
-		std::cerr << " expected 0, got " << nnz( dst ) << "\n";
-		return FAILED;
+	if( nonblocking_execution ) {
+		ret = ret ? ret : grb::clear( dst );
+	} else {
+		if( nnz( dst ) != 0 ) {
+			std::cerr << " expected 0, got " << nnz( dst ) << "\n";
+			return FAILED;
+		}
 	}
 
 	std::cerr << "\b 4:";
 	ret = grb::set< descriptors::dense >( dst, full_mask, 1.0 );
+	ret = ret ? ret : grb::wait( dst );
 	if( ret != ILLEGAL ) {
 		std::cerr << " expected ILLEGAL, got " << toString( ret ) << "\n";
 		return FAILED;
 	}
-	if( nnz( dst ) != 0 ) {
-		std::cerr << " expected 0, got " << nnz( dst ) << "\n";
-		return FAILED;
+	if( nonblocking_execution ) {
+		ret = ret ? ret : grb::clear( dst );
+	} else {
+		if( nnz( dst ) != 0 ) {
+			std::cerr << " expected 0, got " << nnz( dst ) << "\n";
+			return FAILED;
+		}
 	}
 
 	std::cerr << "\b 5:";
 	ret = grb::set< descriptors::dense >( dst, src );
+	ret = ret ? ret : grb::wait( dst );
 	if( ret != ILLEGAL ) {
 		std::cerr << " expected ILLEGAL, got " << toString( ret ) << "\n";
 		return FAILED;
 	}
-	if( nnz( dst ) != 0 ) {
-		std::cerr << " expected 0, got " << nnz( dst ) << "\n";
-		return FAILED;
+	if( nonblocking_execution ) {
+		ret = ret ? ret : grb::clear( dst );
+	} else {
+		if( nnz( dst ) != 0 ) {
+			std::cerr << " expected 0, got " << nnz( dst ) << "\n";
+			return FAILED;
+		}
 	}
 
 	std::cerr << "\b 6:";
 	ret = grb::set< descriptors::dense >( dst, one_mask, src );
+	ret = ret ? ret : grb::wait( dst );
 	if( ret != ILLEGAL ) {
 		std::cerr << " expected ILLEGAL, got " << toString( ret ) << "\n";
 		return FAILED;
 	}
-	if( nnz( dst ) != 0 ) {
-		std::cerr << " expected 0, got " << nnz( dst ) << "\n";
-		return FAILED;
+	if( nonblocking_execution ) {
+		ret = ret ? ret : grb::clear( dst );
```

Comment:
What if grb::clear here does not return success? The test now would not catch such erroneous behaviour is it not?